### PR TITLE
Added support for showing MQ deleted events in change log

### PR DESCRIPTION
**Pull request by hortha**
### Context

Users need to be able to see that an MQ has been deleted from a teaching record and the details of that MQ.

### Changes proposed in this pull request

Update the change log to query MandatoryQualifictionDeletedEvents for the current person and render each entry within the timeline according to the Figma designs.

### Guidance to review

Changes to UI + tests

### Checklist

-   [x] Attach to Trello card
-   [x] Rebased master
-   [x] Cleaned commit history
-   [x] Tested by running locally

### CODE REVIEW

**Comment by hortha** [comment]
[Show MQ deleted events in the change log](https://trello.com/c/tGjNUV8x/1565-show-mq-deleted-events-in-the-change-log)

**Comment by hortha** [comment]
[Show MQ deleted events in the change log](https://trello.com/c/tGjNUV8x/1565-show-mq-deleted-events-in-the-change-log)

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/ChangeLog.cshtml.cs [hunk 1]
```diff
@@ -40,6 +46,47 @@ public async Task<IActionResult> OnGet()
 
         var notesResult = await crmQueryDispatcher.ExecuteQuery(new GetNotesByContactIdQuery(PersonId));
 
+        using var dbContext = await dbContextFactory.CreateDbContextAsync();
+        var personIdString = PersonId.ToString();
+        var eventsWithUser = await dbContext.Database
+            .SqlQuery<EventWithUser>($@"
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/ChangeLog.cshtml.cs:52]: Nit: could use `"""` here to fix the weird indenting

Comment (hortha) [reply TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/ChangeLog.cshtml.cs:52]: ah raw string literals - nice!

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/ChangeLog.cshtml.cs [hunk 2]
```diff
@@ -53,9 +100,54 @@ public async Task<IActionResult> OnGet()
...
 
         return Page();
     }
+
+    public TimelineItem MapTimelineEvent(EventWithUser eventWithUser)
+    {
+        var assembly = typeof(EventBase).Assembly;
+        var eventTypeName = typeof(EventBase).Namespace + "." + eventWithUser.EventName;
+        var eventType = assembly.GetType(eventTypeName);
+        var @event = JsonSerializer.Deserialize(eventWithUser.EventPayload, eventType!);
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/ChangeLog.cshtml.cs:115]: I think should move this into a method on `EventBase` with a corresponding `Serialize` method to centralise it. Can then call it from `Event.FromEventBase` and `Event.ToEventBase` too. Using the `JsonSerializerOptions` on `EventBase` will sort out `RaisedBy` correctly too.
```cs
public static EventBase Deserialize(string payload, string eventName) ...

public string Serialize()...
```

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/EventWithUser.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+namespace TeachingRecordSystem.SupportUi.Pages.Persons.PersonDetail.Timeline.Events;
+
+/// <summary>
+/// Flattened out record to allow Event, TRS User and DQT User to be returned in a single SQL query
+/// </summary>
+public record EventWithUser
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/EventWithUser.cs:6]: Could this be `private` within `ChangeLog.cshtml.cs`?

File: TeachingRecordSystem/tests/TeachingRecordSystem.TestCommon/TestData.UpdateQualification.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
+using Microsoft.Xrm.Sdk.Messages;
+using TeachingRecordSystem.Core.Dqt.Models;
+
+namespace TeachingRecordSystem.TestCommon;
+
+public partial class TestData
+{
+    public Task UpdateQualification(Action<UpdateQualificationBuilder>? configure)
```
Comment (gunndabad) [inline TeachingRecordSystem/tests/TeachingRecordSystem.TestCommon/TestData.UpdateQualification.cs:8]: Can we make this `DeleteMandatoryQualification(Guid qualificationId, ..., bool? syncEnabled = null)` instead (without a builder)?

File: TeachingRecordSystem/tests/TeachingRecordSystem.SupportUi.Tests/PageTests/Persons/PersonDetail/ChangeLogTests.cs [hunk 5]
```diff
@@ -42,10 +45,10 @@ public async Task Get_WithPersonIdForPersonWithNoNotes_DisplaysNoChanges()
     }
 
     [Fact]
-    public async Task Get_WithPersonIdForPersonWithNotes_DisplaysChangesInDescendingOrder()
+    public async Task Get_WithPersonIdForPersonWithChanges_DisplaysChangesInDescendingOrder()
```
Comment (gunndabad) [inline TeachingRecordSystem/tests/TeachingRecordSystem.SupportUi.Tests/PageTests/Persons/PersonDetail/ChangeLogTests.cs:48]: I think we should split this test up and have one for each type of item/event.

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/RaisedByUser.cs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+using TeachingRecordSystem.Core.DataStore.Postgres.Models;
+
+namespace TeachingRecordSystem.SupportUi.Pages.Persons.PersonDetail.Timeline.Events;
+
+public record RaisedByUser
+{
+    public required User? TrsUser { get; set; }
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/RaisedByUser.cs:7]: In `RaisedByUserInfo` we've not used a `Trs` prefix
```suggestion
    public required User? User { get; set; }
```

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/EventWithUser.cs [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+{
+    public required Guid EventId { get; init; }
+    public required string EventName { get; init; }
+    public required DateTime EventCreated { get; init; }
+    public required string EventPayload { get; init; }
+    public required bool EventPublished { get; set; }
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/EventWithUser.cs:12]: Do we need `EventId`, `EventCreated` or `EventPublished`?

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/ChangeLog.cshtml.cs (hunk 2 above)
Comment (hortha) [reply TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/ChangeLog.cshtml.cs:115]: Good idea .. doing that now

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/EventWithUser.cs (hunk 7 above)
Comment (hortha) [reply TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/EventWithUser.cs]: Not how it's used currently we don't

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/ChangeLog.cshtml.cs (hunk 2 above)
Comment (hortha) [reply TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/ChangeLog.cshtml.cs]: Done

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/EventWithUser.cs (hunk 3 above)
Comment (hortha) [reply TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/EventWithUser.cs]: Done

File: TeachingRecordSystem/tests/TeachingRecordSystem.TestCommon/TestData.UpdateQualification.cs (hunk 4 above)
Comment (hortha) [reply TeachingRecordSystem/tests/TeachingRecordSystem.TestCommon/TestData.UpdateQualification.cs]: Done

File: TeachingRecordSystem/tests/TeachingRecordSystem.SupportUi.Tests/PageTests/Persons/PersonDetail/ChangeLogTests.cs (hunk 5 above)
Comment (hortha) [reply TeachingRecordSystem/tests/TeachingRecordSystem.SupportUi.Tests/PageTests/Persons/PersonDetail/ChangeLogTests.cs]: Done

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/RaisedByUser.cs (hunk 6 above)
Comment (hortha) [reply TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/RaisedByUser.cs]: Done

File: TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/EventWithUser.cs (hunk 7 above)
Comment (hortha) [reply TeachingRecordSystem/src/TeachingRecordSystem.SupportUi/Pages/Persons/PersonDetail/Timeline/Events/EventWithUser.cs]: Removed

**Review by gunndabad — APPROVED** [review]

**Pull request merged**